[GStreamer] Use gst_device_reconfigure_element() to configure audio outputs

aperezdc opened the pull request:
#### f22f095d4f9c6b863ce805bfb8e6faf9adf80efc
<pre>
[GStreamer] Use gst_device_reconfigure_element() to configure audio outputs
<a href="https://bugs.webkit.org/show_bug.cgi?id=297418">https://bugs.webkit.org/show_bug.cgi?id=297418</a>

Reviewed by Philippe Normand.

Instead of relying on knowledge about how different GStreamer elements
handle the choice of output device, use gst_device_reconfigure_element()
instead.

This also makes applyAudioSinkDevice() function work with elements which
are not bins. Previously it assumed that the passed element would always
be a bin, which would prevent switching audio outputs correctly when
MediaPlayerPrivateGStreamer::audioSink() returns a non-bin element.

While at it, sprinkle some debug logging and turn applyAudioSinkDevice()
into a method to allow doing so.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::applyAudioSinkDevice):
(WebCore::MediaPlayerPrivateGStreamer::audioOutputDeviceChanged):
(WebCore::applyAudioSinkDevice): Deleted.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/298747@main">https://commits.webkit.org/298747@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c093263f8ada4af4baef8c9720761a3779a1013

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116559 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36220 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26773 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122618 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67123 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118448 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36914 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44808 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88500 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119508 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29436 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104554 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68944 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28501 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22663 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66298 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/98813 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22821 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125765 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43453 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32631 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/97197 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43818 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100763 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96992 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24688 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42304 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20222 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43341 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48932 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42807 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46147 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44512 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -4581,17 +4582,26 @@ void MediaPlayerPrivateGStreamer::checkPlayingConsistency()
+        GST_WARNING_OBJECT(pipeline(), "Skipped unexpected fakesink, your audio configuration may have issues.");
+        return changed;
+    }
+
```
`fakeaudiosink` might need to be handled here too, it's upranked on post-commit bots running layout tests.

aperezdc replied:
Do you have any suggestion about how to handle it? Maybe we can keep this code for non-developer builds, and for developer builds consider that the audio output was done if we reach a fake sink?

aperezdc replied:
Also I think the condition here needs to be changed check for `fakeaudiosink` (instead of `fakesink` which AFAICS is for video).

philn replied:
> Do you have any suggestion about how to handle it? 

Maybe just a GST_DEBUG and return early is sufficient.

> Also I think the condition here needs to be changed check for fakeaudiosink (instead of fakesink which AFAICS is for video).

I don't think so, `autoaudiosink` and `autovideosink` create a `fakesink` as fallback (see `gst_auto_detect_create_fake_element_default()`), so the existing check on `fakesink` looks appropriate.

philn replied:
Yes, because as mentioned above, `autoaudiosink` falls back to `fakesink` if no other sink was found.

aperezdc replied:
Done, thanks again!

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b7f44ee6d20464bd77da3a2a93dea5a3d26954d5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49411

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6c093263f8ada4af4baef8c9720761a3779a1013. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49411

aperezdc committed:
[GStreamer] Use gst_device_reconfigure_element() to configure audio outputs
https://bugs.webkit.org/show_bug.cgi?id=297418

Reviewed by Philippe Normand.

Instead of relying on knowledge about how different GStreamer elements
handle the choice of output device, use gst_device_reconfigure_element()
instead.

This also makes applyAudioSinkDevice() function work with elements which
are not bins. Previously it assumed that the passed element would always
be a bin, which would prevent switching audio outputs correctly when
MediaPlayerPrivateGStreamer::audioSink() returns a non-bin element.

While at it, sprinkle some debug logging and turn applyAudioSinkDevice()
into a method to allow doing so.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::applyAudioSinkDevice):
(WebCore::MediaPlayerPrivateGStreamer::audioOutputDeviceChanged):
(WebCore::applyAudioSinkDevice): Deleted.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:

Canonical link: https://commits.webkit.org/298747@main
